### DISCUSSION TITLE
Newbie guide to A/B kernel/dtb updating and fallback

### DISCUSSION THREAD

**Discussion by arrow53** [OE4T/tegra-demo-distro#292] [Q&A] [answered]
I have a stock orin nano dev kit using kirkstone with machine `jetson-orin-nano-devkit-nvme`

I'd like to be able to manually update my bootloader stuff, namely kernel/dtb. I'm kinda fuzzy on the procedure / tools. I think it goes something like this but I'm missing some steps that I'm hoping I can get help filling in.

1. I added this to my `local.conf`
```
IMAGE_INSTALL:append = " tegra-bup-payload"
TNSPEC_BOOTDEV = "nvme0n1p1"
```
2. I can boot the orin and `nvbootctrl get-number-slots` returns 2 slots
3. I see a file in my deploy directory `tegra-minimal-initramfs-jetson-orin-nano-devkit-nvme.bl_only.bup-payload` that I copy over to my target into `/opt/ota_package/`

But, now I'm lost. Do I use `nv_update_engine`? This seems to want to update both the A/B slots. I'd like to just update the non-active slot, make sure that finishes, then manually switch to that slot.

thanks!

**Comment by kekiefer** [comment, marked as answer]
The modern approach to bootloader updates is the capsule update. You deploy a special payload into to the esp partition, tell uefi about it, and it handles the rest.

Basically:
* build `tegra-uefi-capsules`
* install the `tegra-bl.cap` to `/opt/nvidia/esp/EFI/UpdateCapsule/TEGRA_BL.Cap`, creating the destination directory if needed
* execute `/usr/bin/oe4t-set-uefi-OSIndications` to tell uefi there's an update available
* reboot

If you use swupdate (or even if you don't), here is a good example that shows where the capsule package goes, and how to flag down UEFI: https://github.com/Trellis-Logic/meta-swupdate-boards/pull/1/files

**Reply by arrow53** [reply]
@kekiefer wow, this is perfect. Thanks a ton for pointing me in this direction.

**Reply by arrow53** [reply]
@kekiefer is the `tegra-bl.cap` just the bootloader? When I modify my device tree and try to update using this method I don't see a change. I see in the recipe there is also a `tegra-kernel.cap`. Is that what I want instead? If so, how do I generate it?

**Reply by kekiefer** [reply]
It looks like tegra-kernel.cap is generated at the same time, and put in the deploy dir the same way, so you should be able to use it. In the past we had different update engine payloads for kernel, bootloader, or everything.

I use extlinux with my own setup, so I don't have personal experience trying to add both payloads, and I don't know if it will work or not. Maybe someone else has tried and can provide some input here, or you could just try it and report back.

**Reply by arrow53** [reply]
ok. I'll do some digging. I'm not getting a `tegra-kernel.cap` so unclear why that's not happening.
```
build/tmp/deploy/images/jetson-orin-nano-devkit-nvme$ ls *cap
jetson-orin-nano-devkit-nvme-tegra-bl.cap  tegra-bl.cap
```
thanks.

**Reply by arrow53** [reply]
@kekiefer seems capsules only work for non external partitions. Orin nano / nx have dtb / kernel on external drive as why I'm probably not getting the `tegra-kernel.cap`

**Reply by minicarpet** [reply]
Hi, I have 2 questions:

- why the file should be copied into `/opt/nvidia/esp/EFI/UpdateCapsule/TEGRA_BL.Cap` ? Were can I found the NVIDIA documentation about this ?
- I didn't find any documentation of `oe4t-set-uefi-OSIndications`, were can I find it ? It seems to never print an error and always return success.

**Reply by minicarpet** [reply]
I've just found [this](https://docs.nvidia.com/jetson/archives/r36.4.4/DeveloperGuide/SD/Bootloader/UpdateAndRedundancy.html#manually-trigger-the-capsule-update) what I am not sure is, can I mount the esp partition (which is /boot folder right ?) and bind it using fstab ?
`/boot/efi/EFI/UpdateCapsule  /opt/nvidia/esp/EFI/UpdateCapsule       none      bind,rw,nofail,x-systemd.requires-mounts-for=/boot/efi                     0  0`
Indeed if rootfs is read-only I cannot create  `/opt/nvidia/esp/EFI/UpdateCapsule/TEGRA_BL.Cap` so how handle this case ?

**Reply by dwalkes** [reply]
> can I mount the esp partition (which is /boot folder right ?) and bind it using fstab ?

It should be mounted for you by [setup-nv-boot-control](https://github.com/OE4T/meta-tegra/blob/master/recipes-bsp/tools/setup-nv-boot-control/esp.mount.in) as long as you have the necessary yocto dependencies included in your image.

If you are looking for a turn key example build which supports demonstrating use of capsules for update see https://github.com/OE4T/tegra-demo-distro/tree/master/layers/meta-tegrademo/dynamic-layers/meta-swupdate.

**Reply by minicarpet** [reply]
It mounts to /boot/efi, right ? Sooo why not copying the TEGRA_BL.Cap file into /boot/efi (ie /boot/efi/EFI/UpdateCapsule/TEGRA_BL.Cap) directly from swupdate ? Do I miss a part about the bootloader update ?

**Reply by dwalkes** [reply]
That should be exactly what it's doing at https://github.com/OE4T/tegra-demo-distro/blob/698859624de92b3987e1be210820412ff9c603b0/layers/meta-tegrademo/dynamic-layers/meta-swupdate/recipes-support/swupdate/tegra-swupdate.inc#L33 and https://github.com/OE4T/tegra-demo-distro/blob/698859624de92b3987e1be210820412ff9c603b0/layers/meta-tegrademo/dynamic-layers/meta-swupdate/recipes-demo/images/swupdate-image-tegra/sw-description#L41-L48

**Reply by minicarpet** [reply]
That perfectly clear for me now! Thank you !

**Comment by arrow53** [comment]
@dwalkes I read this thread [element](https://app.element.io/#/room/#OE4T_jetpack-5-device-update:gitter.im) but i'm a bit confused

For an orin nano with external NVME is the recommended way to update dtb / kernel to use meta-swupdate? I'm not sure I can use this layer for my purposes.

If I wanted to take a more manual approach could I just mount the dtb / kernel partitions and copy new files in place?

**Reply by kekiefer** [reply]
Your simplest path might be to just to use extlinux, then the kernel and dtb are located within the rootfs for update.

I'm not sure if you can simply just update the partitions -- at least in the past they had some signature applied to them.

**Reply by arrow53** [reply]
@kekiefer I mean that sounds good to me. Ideally I'd be able to just manually put in place dtb, kernel, and my own rootfs updates in A / B partitions.

Ok, dumb question, how do I go from UEFI to extlinux?

**Reply by arrow53** [reply]
reading more about it, it seems better to stick with UEFI and try to figure out how to get it to work with that as UEFI is the future.

**Reply by kekiefer** [reply]
In our case, "extlinux" refers to a boot mode for the same UEFI loader, as opposed to the "direct" mode. See here: https://github.com/OE4T/meta-tegra/wiki/extlinux.conf-support

**Reply by arrow53** [reply]
@kekiefer very interesting. It seems this won't work with stock A/B redudancy?

https://github.com/OE4T/meta-tegra/wiki/extlinux.conf-support#incompatible-with-ab-redundancy

I don't see the `cboot-extlinux` package at any rate. This may have gotten nix'd in later releases

https://github.com/OE4T/meta-tegra/blob/nanbield/classes/l4t-extlinux-config.bbclass

**Reply by madisongh** [reply]
For R35.x and later, the recipe for creating the file is `l4t-launcher-extlinux.bb`.
